File: PrimeMinistersByJava/primeministers/Downloader.java
package primeministers;

import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.List;

import javax.imageio.ImageIO;

import libraries.Condition;


public class Downloader extends IO {

	/**
	* CSVの参照先であるベースのURL
	*/
	// private URL url;
	private String url;

	/**
	* ダウンロードを行うためのコンストラクタ
	* @params aClassName Enumから与えられたクラス名
	*/
	public Downloader(Table aTable) {
		super(aTable);
		String aClassName = aTable.aClassAttribute().toString();
		this.url = String.format("%s%s/%s.csv",super.table().attributes().baseUrl(), aClassName, aClassName);
	}

	/**
	* CSVをダウンロードする
	* @return File CSVの保存先
	*/
	public File downloadCsv() {
		super.table().attributes().setBaseDirectory();
		File aTargetFilepath = new File(Attributes.getBaseDirectory(), Attributes.className().concat(".csv"));
		this.downloadFile(this.url.toString(), aTargetFilepath.getPath());
		return aTargetFilepath;
	}

	/**
	* ダウンロードを統括するメソッド
	* @params String aTargetURL 保存するURL
	* @params String aTargetFilepath 保存先のファイルパス
	*/
	private void downloadFile(String aTargetURL, String aTargetFilepath) {
		List<String> listOfTargetURL = super.readTextFromURL(aTargetURL);
		File aCSVFilePath = super.writeText(listOfTargetURL, aTargetFilepath);
		return;
	}

	/**
	* 画像群をダウンロードする
	*/
	public void downloadImages() {
		this.downloadPictures(super.attributes().indexOfImage());
		return;
	}

	/**
	* 画像群またはサムネイル画像群をダウンロードする
	*/
	public void downloadPictures(final int indexOfPictures) {
		new Condition(() -> super.tuples() == null)
		.ifThenElse(() -> {
			System.out.println("Tuple is null.");
		},
		() -> {
			System.out.println("Yes, tuple.");
			super.tuples().stream()
			.map(aTuple->aTuple.values().get(indexOfPictures))
			.forEach(item -> {
				File aFile = new File(super.table().attributes().baseDirectory(), item);
				String directoryOfItem = aFile.getPath();

				String urlOfItem = super.table().attributes().baseUrl().concat(
					new File(
					Attributes.className(), item).getPath()
				);
				System.out.printf("[Downloading...] %s -> %s%n", urlOfItem, directoryOfItem);
				BufferedImage anBufferedImage = IO.readImageFromURL(urlOfItem);
				super.writeImage(anBufferedImage, directoryOfItem);
				System.out.printf("complete!!%n%n");
			});
		});
	}

	/**
	* サムネイル画像群をダウンロードする
	*/
	public void downloadThumbnails() {
		this.downloadPictures(super.attributes().indexOfThumbnail());
		return;
	}

	public void perform() {
		// プロジェクトルートディレクトリの作成
		File desktopPath = new File(System.getProperty("user.home"), "Desktop");
		File aDestinationDirectory = new File(desktopPath, super.table().aClassAttribute().toString().concat("_Java")); // デスクトップにディレクトリ作成
		aDestinationDirectory.mkdir();
		new File(aDestinationDirectory, "images").mkdir();
		new File(aDestinationDirectory, "thumbnails").mkdir();
		// ダウンロード作業
		AtomicInteger anAtomicInteger = new AtomicInteger();
		File aCSVFilePath = this.downloadCsv();
		final Reader aReader = new Reader(super.table(), aCSVFilePath, super.table().aClassAttribute());
		List<List<String>> splittedList = aReader.perform();
		splittedList.forEach(aLine -> {
			new Condition(() -> anAtomicInteger.getAndIncrement() != 0).
			ifTrue(() -> {
				Tuple aTuple = new Tuple(aLine);
				super.table().add(new Tuple(aLine));
			});
		});
		this.downloadImages();
		this.downloadThumbnails();
	}

}

File: PrimeMinistersByPython/primeministers/downloader.py
from .attributes import Attributes
from .io import IO
from .table import Table
from .reader import Reader
from .tuple import Tuple

import os
import sys
sys.path.append('../')
from libraries.condition import Condition

class Downloader(IO):
	def __init__(self, a_table):
		super(Downloader, self).__init__(a_table)
		a_classname = a_table.get_a_class_attribute()
		self.url = "{0}{1}/{2}.csv".format(self.get_table().get_attributes().base_url(), a_classname, a_classname)
		return

	def download_csv(self):
		self.get_table().get_attributes().set_base_directory()
		a_string = self.get_table().get_attributes().get_base_directory()
		a_target_filepath = '{0}/{1}.csv'.format(a_string, a_string.split('/')[-1])
		self.download_file(self.url, a_target_filepath)
		return a_target_filepath

	def download_file(self, a_target_url, a_target_filepath):
		list_of_target_url = self.read_text_from_url(a_target_url)
		a_csv_filepath = self.write_text(list_of_target_url, a_target_filepath)
		return

	def download_images(self):
		self.download_pictures(self.get_table().get_attributes().index_of_image())
		return

	def download_pictures(self, index_of_pictures):
		Condition(self.get_tuples() == None).if_then_else(
			lambda: print('Tuple is None.'),
			lambda: self.download_pictures_from_url(index_of_pictures)
		)
		return

	def download_pictures_from_url(self, index_of_pictures):
		print('Yes, tuple.')
		for a_string in map(lambda item: item.get_values_with_index(index_of_pictures), self.get_tuples()):
			a_file = os.path.join(Attributes.get_base_directory(), a_string)
			url_of_item = os.path.join('{0}{1}'.format(self.get_table().get_attributes().base_url(), Attributes.get_class_name()), a_string)
			print('[Downloading...] {0} -> {1}'.format(url_of_item, a_file))
			self.write_image(url_of_item, a_file)
			print('complete!!\n')
		return

	def download_thumbnails(self):
		self.download_pictures(self.get_table().get_attributes().index_of_thumbnail())
		return

	def perform(self):
		import os
		a_desktop_path = Attributes.desktop_path()
		a_destination_directory = os.path.join(a_desktop_path, "{0}_Python".format(self.get_table().get_a_class_attribute()))
		os.makedirs(a_destination_directory, exist_ok=True)
		os.makedirs(os.path.join(a_destination_directory, "images"), exist_ok=True)
		os.makedirs(os.path.join(a_destination_directory, "thumbnails"), exist_ok=True)
		a_csv_filepath = self.download_csv()
		a_reader = Reader(self.get_table(), a_csv_filepath, self.get_table().get_a_class_attribute())
		# ここから続き
		for an_index, a_line in enumerate(a_reader.perform()):
			Condition(an_index != 0).if_true(lambda: self.get_table().add(Tuple(a_line)))
		self.download_images()
		self.download_thumbnails()
		return self.get_tuples()
